### PR TITLE
Support `e4m3fn` KV cache

### PR DESCRIPTION
**Pull request by danieldk**
# What does this PR do?

Add support for `e4m3fn` KV caches as well.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: server/text_generation_server/layers/attention/kv_cache.py [hunk 1]
```diff
@@ -24,11 +24,13 @@ def __init__(
     ):
         """Construct the key-value cache for a layer."""
 
-        if dtype == torch.float8_e5m2 and (
-            ATTENTION != "flashinfer" or SYSTEM != "cuda"
+        if (
+            dtype.itemsize == 1
+            and dtype.is_floating_point
```
Comment (Narsil) [inline server/text_generation_server/layers/attention/kv_cache.py:29]: Why not `in {torch.float8_e5m2, torch.float8_e4m3}` ?

Comment (danieldk) [reply server/text_generation_server/layers/attention/kv_cache.py:29]: We might have other variants in the future as well (`e4m3fnuz`, `e5m2uz`).

Comment (Narsil) [reply server/text_generation_server/layers/attention/kv_cache.py:29]: Fair enough. From my understanding they are actually the same dtypes only side effectifs on edge values (which we won't check against).

I don't even think torch have plans for explicit support https://pytorch.org/docs/stable/tensors.html#data-types

As you wish for merging, I just feel like having consistent checks everywhere will lead to fewer bugs.

Comment (danieldk) [reply server/text_generation_server/layers/attention/kv_cache.py:29]: Yeah, that's a fair point. I think `in {torch.float8_e5m2, torch.float8_e4m3}` makes it easier to grep for all the places where one of these float types is used. So I'll update the PR.

**Pull request merged**